src/services/textChanges.ts

```diff
@@ -916,7 +927,7 @@ namespace ts.textChanges {
                 const normalized = stableSort(changesInFile, (a, b) => (a.range.pos - b.range.pos) || (a.range.end - b.range.end));
                 // verify that change intervals do not overlap, except possibly at end points.
                 for (let i = 0; i < normalized.length - 1; i++) {
-                    Debug.assert(normalized[i].range.end <= normalized[i + 1].range.pos, "Changes overlap", () =>
+                    Debug.assert(normalized[i].range.end <= normalized[i + 1].range.pos || normalized[i].kind === ChangeKind.ReplaceWithList, "Changes overlap", () =>

```

Comment:
Could you paste the two JSON/debugger output of the two changes (`normalized[i]` and `normalized[i + 1]`) that trigger this assertion? I have some thoughts, but want to make sure I understand what’s happening correctly.

Reply:
They are pretty big nodes. But here it goes:
```
normalized[i] = {
  "kind": 4,
  "sourceFile": {
    "pos": 0,
    "end": 133,
    "flags": 0,
    "modifierFlagsCache": 0,
    "transformFlags": 1,
    "kind": 298,
    "statements": [
      {
        "pos": 0,
        "end": 39,
        "flags": 0,
        "modifierFlagsCache": 536870912,
        "transformFlags": 0,
        "kind": 262,
        "importClause": {
          "pos": 6,
          "end": 23,
          "flags": 0,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "kind": 263,
          "isTypeOnly": false,
          "namedBindings": {
            "pos": 6,
            "end": 23,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 265,
            "elements": [
              {
                "pos": 8,
                "end": 14,
                "flags": 0,
                "modifierFlagsCache": 536870912,
                "transformFlags": 0,
                "kind": 266,
                "name": {
                  "pos": 8,
                  "end": 14,
                  "flags": 0,
                  "modifierFlagsCache": 0,
                  "transformFlags": 0,
                  "kind": 78,
                  "escapedText": "Test1",
                  "flowNode": {
                    "flags": 2
                  }
                },
                "symbol": {
                  "flags": 2097152,
                  "escapedName": "Test1",
                  "declarations": [
                    null
                  ],
                  "id": 14,
                  "isReferenced": 788968
                }
              },
              {
                "pos": 15,
                "end": 21,
                "flags": 0,
                "modifierFlagsCache": 536870912,
                "transformFlags": 0,
                "kind": 266,
                "name": {
                  "pos": 15,
                  "end": 21,
                  "flags": 0,
                  "modifierFlagsCache": 0,
                  "transformFlags": 0,
                  "kind": 78,
                  "escapedText": "Test4"
                },
                "symbol": {
                  "flags": 2097152,
                  "escapedName": "Test4",
                  "declarations": [
                    null
                  ],
                  "id": 17,
                  "isReferenced": 788968
                },
                "_children": [
                  null
                ]
              }
            ],
            "_children": [
              {
                "pos": 6,
                "end": 8,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 18
              },
              {
                "pos": 8,
                "end": 21,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 338,
                "_children": [
                  null,
                  {
                    "pos": 14,
                    "end": 15,
                    "flags": 0,
                    "modifierFlagsCache": 0,
                    "transformFlags": 0,
                    "kind": 27
                  },
                  null
                ]
              },
              {
                "pos": 21,
                "end": 23,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 19
              }
            ]
          },
          "_children": [
            null
          ],
          "id": 16
        },
        "moduleSpecifier": {
          "pos": 28,
          "end": 38,
          "flags": 0,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "kind": 10,
          "text": "./file1",
          "hasExtendedUnicodeEscape": false
        },
        "_children": [
          {
            "pos": 0,
            "end": 6,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 99
          },
          null,
          {
            "pos": 23,
            "end": 28,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 153
          },
          null,
          {
            "pos": 38,
            "end": 39,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 26
          }
        ]
      },
      {
        "pos": 39,
        "end": 133,
        "flags": 524288,
        "modifierFlagsCache": 536870912,
        "transformFlags": 1,
        "kind": 254,
        "symbol": {
          "flags": 64,
          "escapedName": "Testing",
          "declarations": [
            null
          ],
          "members": {},
          "id": 19
        },
        "name": {
          "pos": 49,
          "end": 57,
          "flags": 524288,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "kind": 78,
          "escapedText": "Testing"
        },
        "members": [
          {
            "pos": 59,
            "end": 77,
            "flags": 524288,
            "modifierFlagsCache": 536870912,
            "transformFlags": 1,
            "kind": 163,
            "symbol": {
              "flags": 4,
              "escapedName": "test1",
              "declarations": [
                null
              ],
              "id": 20
            },
            "name": {
              "pos": 59,
              "end": 69,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 0,
              "kind": 78,
              "escapedText": "test1"
            },
            "type": {
              "pos": 70,
              "end": 76,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 1,
              "kind": 174,
              "typeName": {
                "pos": 70,
                "end": 76,
                "flags": 524288,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 78,
                "escapedText": "Test1"
              },
              "id": 12
            }
          },
          {
            "pos": 77,
            "end": 95,
            "flags": 524288,
            "modifierFlagsCache": 536870912,
            "transformFlags": 1,
            "kind": 163,
            "symbol": {
              "flags": 4,
              "escapedName": "test2",
              "declarations": [
                null
              ],
              "id": 21
            },
            "name": {
              "pos": 77,
              "end": 87,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 0,
              "kind": 78,
              "escapedText": "test2"
            },
            "type": {
              "pos": 88,
              "end": 94,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 1,
              "kind": 174,
              "typeName": {
                "pos": 88,
                "end": 94,
                "flags": 524288,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 78,
                "escapedText": "Test2"
              },
              "id": 13,
              "_children": [
                null
              ]
            },
            "_children": [
              null,
              {
                "pos": 87,
                "end": 88,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 58
              },
              null,
              {
                "pos": 94,
                "end": 95,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 26
              }
            ]
          },
          {
            "pos": 95,
            "end": 113,
            "flags": 524288,
            "modifierFlagsCache": 536870912,
            "transformFlags": 1,
            "kind": 163,
            "symbol": {
              "flags": 4,
              "escapedName": "test3",
              "declarations": [
                null
              ],
              "id": 22
            },
            "name": {
              "pos": 95,
              "end": 105,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 0,
              "kind": 78,
              "escapedText": "test3"
            },
            "type": {
              "pos": 106,
              "end": 112,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 1,
              "kind": 174,
              "typeName": {
                "pos": 106,
                "end": 112,
                "flags": 524288,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 78,
                "escapedText": "Test3"
              },
              "id": 14,
              "_children": [
                null
              ]
            },
            "_children": [
              null,
              {
                "pos": 105,
                "end": 106,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 58
              },
              null,
              {
                "pos": 112,
                "end": 113,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 26
              }
            ]
          },
          {
            "pos": 113,
            "end": 131,
            "flags": 524288,
            "modifierFlagsCache": 536870912,
            "transformFlags": 1,
            "kind": 163,
            "symbol": {
              "flags": 4,
              "escapedName": "test4",
              "declarations": [
                null
              ],
              "id": 23
            },
            "name": {
              "pos": 113,
              "end": 123,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 0,
              "kind": 78,
              "escapedText": "test4"
            },
            "type": {
              "pos": 124,
              "end": 130,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 1,
              "kind": 174,
              "typeName": {
                "pos": 124,
                "end": 130,
                "flags": 524288,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 78,
                "escapedText": "Test4"
              },
              "id": 15
            }
          }
        ],
        "_children": [
          {
            "pos": 39,
            "end": 49,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 117
          },
          null,
          {
            "pos": 57,
            "end": 59,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 18
          },
          {
            "pos": 59,
            "end": 131,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 338,
            "_children": [
              null,
              null,
              null,
              null
            ]
          },
          {
            "pos": 131,
            "end": 133,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 19
          }
        ]
      }
    ],
    "endOfFileToken": {
      "pos": 133,
      "end": 133,
      "flags": 0,
      "modifierFlagsCache": 0,
      "transformFlags": 0,
      "kind": 1
    },
    "fileName": "/tests/cases/fourslash/file2.ts",
    "text": "import { Test1, Test4 } from './file1';\ninterface Testing {\n    test1: Test1;\n    test2: Test2;\n    test3: Test3;\n    test4: Test4;\n}",
    "languageVersion": 1,
    "languageVariant": 0,
    "scriptKind": 3,
    "isDeclarationFile": false,
    "hasNoDefaultLib": false,
    "bindDiagnostics": [],
    "pragmas": {},
    "referencedFiles": [],
    "typeReferenceDirectives": [],
    "libReferenceDirectives": [],
    "amdDependencies": [],
    "nodeCount": 28,
    "identifierCount": 11,
    "identifiers": {},
    "parseDiagnostics": [],
    "version": "1",
    "scriptSnapshot": {
      "scriptInfo": {
        "fileName": "/tests/cases/fourslash/file2.ts",
        "content": "import { Test1, Test4 } from './file1';\ninterface Testing {\n    test1: Test1;\n    test2: Test2;\n    test3: Test3;\n    test4: Test4;\n}",
        "isRootFile": true,
        "version": 1,
        "editRanges": []
      },
      "textSnapshot": "import { Test1, Test4 } from './file1';\ninterface Testing {\n    test1: Test1;\n    test2: Test2;\n    test3: Test3;\n    test4: Test4;\n}",
      "version": 1
    },
    "path": "/tests/cases/fourslash/file2.ts",
    "resolvedPath": "/tests/cases/fourslash/file2.ts",
    "originalFileName": "/tests/cases/fourslash/file2.ts",
    "imports": [
      null
    ],
    "moduleAugmentations": [],
    "resolvedModules": {},
    "symbol": {
      "flags": 512,
      "escapedName": "\"/tests/cases/fourslash/file2\"",
      "declarations": [
        null
      ],
      "exports": {},
      "id": 24
    },
    "locals": {},
    "symbolCount": 8,
    "classifiableNames": {},
    "id": 11,
    "_children": [
      {
        "pos": 0,
        "end": 133,
        "flags": 0,
        "modifierFlagsCache": 0,
        "transformFlags": 0,
        "kind": 338,
        "_children": [
          null,
          null
        ]
      },
      null
    ],
    "lineMap": [
      0,
      40,
      60,
      78,
      96,
      114,
      132
    ]
  },
  "range": {
    "pos": 15,
    "end": 16
  },
  "options": {
    "prefix": " ",
    "suffix": ", "
  },
  "node": {
    "pos": -1,
    "end": -1,
    "flags": 8,
    "modifierFlagsCache": 0,
    "transformFlags": 0,
    "kind": 266,
    "name": {
      "pos": -1,
      "end": -1,
      "flags": 8,
      "modifierFlagsCache": 0,
      "transformFlags": 0,
      "kind": 78,
      "escapedText": "Test2"
    }
  }
}
```
```
normalized[i + 1] = {
  "kind": 4,
  "sourceFile": {
    "pos": 0,
    "end": 133,
    "flags": 0,
    "modifierFlagsCache": 0,
    "transformFlags": 1,
    "kind": 298,
    "statements": [
      {
        "pos": 0,
        "end": 39,
        "flags": 0,
        "modifierFlagsCache": 536870912,
        "transformFlags": 0,
        "kind": 262,
        "importClause": {
          "pos": 6,
          "end": 23,
          "flags": 0,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "kind": 263,
          "isTypeOnly": false,
          "namedBindings": {
            "pos": 6,
            "end": 23,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 265,
            "elements": [
              {
                "pos": 8,
                "end": 14,
                "flags": 0,
                "modifierFlagsCache": 536870912,
                "transformFlags": 0,
                "kind": 266,
                "name": {
                  "pos": 8,
                  "end": 14,
                  "flags": 0,
                  "modifierFlagsCache": 0,
                  "transformFlags": 0,
                  "kind": 78,
                  "escapedText": "Test1",
                  "flowNode": {
                    "flags": 2
                  }
                },
                "symbol": {
                  "flags": 2097152,
                  "escapedName": "Test1",
                  "declarations": [
                    null
                  ],
                  "id": 14,
                  "isReferenced": 788968
                }
              },
              {
                "pos": 15,
                "end": 21,
                "flags": 0,
                "modifierFlagsCache": 536870912,
                "transformFlags": 0,
                "kind": 266,
                "name": {
                  "pos": 15,
                  "end": 21,
                  "flags": 0,
                  "modifierFlagsCache": 0,
                  "transformFlags": 0,
                  "kind": 78,
                  "escapedText": "Test4"
                },
                "symbol": {
                  "flags": 2097152,
                  "escapedName": "Test4",
                  "declarations": [
                    null
                  ],
                  "id": 17,
                  "isReferenced": 788968
                },
                "_children": [
                  null
                ]
              }
            ],
            "_children": [
              {
                "pos": 6,
                "end": 8,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 18
              },
              {
                "pos": 8,
                "end": 21,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 338,
                "_children": [
                  null,
                  {
                    "pos": 14,
                    "end": 15,
                    "flags": 0,
                    "modifierFlagsCache": 0,
                    "transformFlags": 0,
                    "kind": 27
                  },
                  null
                ]
              },
              {
                "pos": 21,
                "end": 23,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 19
              }
            ]
          },
          "_children": [
            null
          ],
          "id": 16
        },
        "moduleSpecifier": {
          "pos": 28,
          "end": 38,
          "flags": 0,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "kind": 10,
          "text": "./file1",
          "hasExtendedUnicodeEscape": false
        },
        "_children": [
          {
            "pos": 0,
            "end": 6,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 99
          },
          null,
          {
            "pos": 23,
            "end": 28,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 153
          },
          null,
          {
            "pos": 38,
            "end": 39,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 26
          }
        ]
      },
      {
        "pos": 39,
        "end": 133,
        "flags": 524288,
        "modifierFlagsCache": 536870912,
        "transformFlags": 1,
        "kind": 254,
        "symbol": {
          "flags": 64,
          "escapedName": "Testing",
          "declarations": [
            null
          ],
          "members": {},
          "id": 19
        },
        "name": {
          "pos": 49,
          "end": 57,
          "flags": 524288,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "kind": 78,
          "escapedText": "Testing"
        },
        "members": [
          {
            "pos": 59,
            "end": 77,
            "flags": 524288,
            "modifierFlagsCache": 536870912,
            "transformFlags": 1,
            "kind": 163,
            "symbol": {
              "flags": 4,
              "escapedName": "test1",
              "declarations": [
                null
              ],
              "id": 20
            },
            "name": {
              "pos": 59,
              "end": 69,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 0,
              "kind": 78,
              "escapedText": "test1"
            },
            "type": {
              "pos": 70,
              "end": 76,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 1,
              "kind": 174,
              "typeName": {
                "pos": 70,
                "end": 76,
                "flags": 524288,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 78,
                "escapedText": "Test1"
              },
              "id": 12
            }
          },
          {
            "pos": 77,
            "end": 95,
            "flags": 524288,
            "modifierFlagsCache": 536870912,
            "transformFlags": 1,
            "kind": 163,
            "symbol": {
              "flags": 4,
              "escapedName": "test2",
              "declarations": [
                null
              ],
              "id": 21
            },
            "name": {
              "pos": 77,
              "end": 87,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 0,
              "kind": 78,
              "escapedText": "test2"
            },
            "type": {
              "pos": 88,
              "end": 94,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 1,
              "kind": 174,
              "typeName": {
                "pos": 88,
                "end": 94,
                "flags": 524288,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 78,
                "escapedText": "Test2"
              },
              "id": 13,
              "_children": [
                null
              ]
            },
            "_children": [
              null,
              {
                "pos": 87,
                "end": 88,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 58
              },
              null,
              {
                "pos": 94,
                "end": 95,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 26
              }
            ]
          },
          {
            "pos": 95,
            "end": 113,
            "flags": 524288,
            "modifierFlagsCache": 536870912,
            "transformFlags": 1,
            "kind": 163,
            "symbol": {
              "flags": 4,
              "escapedName": "test3",
              "declarations": [
                null
              ],
              "id": 22
            },
            "name": {
              "pos": 95,
              "end": 105,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 0,
              "kind": 78,
              "escapedText": "test3"
            },
            "type": {
              "pos": 106,
              "end": 112,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 1,
              "kind": 174,
              "typeName": {
                "pos": 106,
                "end": 112,
                "flags": 524288,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 78,
                "escapedText": "Test3"
              },
              "id": 14,
              "_children": [
                null
              ]
            },
            "_children": [
              null,
              {
                "pos": 105,
                "end": 106,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 58
              },
              null,
              {
                "pos": 112,
                "end": 113,
                "flags": 0,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 26
              }
            ]
          },
          {
            "pos": 113,
            "end": 131,
            "flags": 524288,
            "modifierFlagsCache": 536870912,
            "transformFlags": 1,
            "kind": 163,
            "symbol": {
              "flags": 4,
              "escapedName": "test4",
              "declarations": [
                null
              ],
              "id": 23
            },
            "name": {
              "pos": 113,
              "end": 123,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 0,
              "kind": 78,
              "escapedText": "test4"
            },
            "type": {
              "pos": 124,
              "end": 130,
              "flags": 524288,
              "modifierFlagsCache": 0,
              "transformFlags": 1,
              "kind": 174,
              "typeName": {
                "pos": 124,
                "end": 130,
                "flags": 524288,
                "modifierFlagsCache": 0,
                "transformFlags": 0,
                "kind": 78,
                "escapedText": "Test4"
              },
              "id": 15
            }
          }
        ],
        "_children": [
          {
            "pos": 39,
            "end": 49,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 117
          },
          null,
          {
            "pos": 57,
            "end": 59,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 18
          },
          {
            "pos": 59,
            "end": 131,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 338,
            "_children": [
              null,
              null,
              null,
              null
            ]
          },
          {
            "pos": 131,
            "end": 133,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "kind": 19
          }
        ]
      }
    ],
    "endOfFileToken": {
      "pos": 133,
      "end": 133,
      "flags": 0,
      "modifierFlagsCache": 0,
      "transformFlags": 0,
      "kind": 1
    },
    "fileName": "/tests/cases/fourslash/file2.ts",
    "text": "import { Test1, Test4 } from './file1';\ninterface Testing {\n    test1: Test1;\n    test2: Test2;\n    test3: Test3;\n    test4: Test4;\n}",
    "languageVersion": 1,
    "languageVariant": 0,
    "scriptKind": 3,
    "isDeclarationFile": false,
    "hasNoDefaultLib": false,
    "bindDiagnostics": [],
    "pragmas": {},
    "referencedFiles": [],
    "typeReferenceDirectives": [],
    "libReferenceDirectives": [],
    "amdDependencies": [],
    "nodeCount": 28,
    "identifierCount": 11,
    "identifiers": {},
    "parseDiagnostics": [],
    "version": "1",
    "scriptSnapshot": {
      "scriptInfo": {
        "fileName": "/tests/cases/fourslash/file2.ts",
        "content": "import { Test1, Test4 } from './file1';\ninterface Testing {\n    test1: Test1;\n    test2: Test2;\n    test3: Test3;\n    test4: Test4;\n}",
        "isRootFile": true,
        "version": 1,
        "editRanges": []
      },
      "textSnapshot": "import { Test1, Test4 } from './file1';\ninterface Testing {\n    test1: Test1;\n    test2: Test2;\n    test3: Test3;\n    test4: Test4;\n}",
      "version": 1
    },
    "path": "/tests/cases/fourslash/file2.ts",
    "resolvedPath": "/tests/cases/fourslash/file2.ts",
    "originalFileName": "/tests/cases/fourslash/file2.ts",
    "imports": [
      null
    ],
    "moduleAugmentations": [],
    "resolvedModules": {},
    "symbol": {
      "flags": 512,
      "escapedName": "\"/tests/cases/fourslash/file2\"",
      "declarations": [
        null
      ],
      "exports": {},
      "id": 24
    },
    "locals": {},
    "symbolCount": 8,
    "classifiableNames": {},
    "id": 11,
    "_children": [
      {
        "pos": 0,
        "end": 133,
        "flags": 0,
        "modifierFlagsCache": 0,
        "transformFlags": 0,
        "kind": 338,
        "_children": [
          null,
          null
        ]
      },
      null
    ],
    "lineMap": [
      0,
      40,
      60,
      78,
      96,
      114,
      132
    ]
  },
  "range": {
    "pos": 15,
    "end": 16
  },
  "options": {
    "prefix": " ",
    "suffix": ", "
  },
  "node": {
    "pos": -1,
    "end": -1,
    "flags": 8,
    "modifierFlagsCache": 0,
    "transformFlags": 0,
    "kind": 266,
    "name": {
      "pos": -1,
      "end": -1,
      "flags": 8,
      "modifierFlagsCache": 0,
      "transformFlags": 0,
      "kind": 78,
      "escapedText": "Test3"
    }
  }
}
```

Reply:
Edited ^ to remove the `sourceFile` node so we can look at this. The source file text was:

```ts
import { Test1, Test4 } from './file1';
interface Testing {
    test1: Test1;
    test2: Test2;
    test3: Test3;
    test4: Test4;
}
```

Reply:
What’s happening is that both fixes are attempting to delete the space between `Test1,` and `Test4` with their inserted node. They make up for the deletion of the space by including it in the `prefix` of their insertion. But it’s the deletion of the space that’s the alleged problem, because overlapping deletes might be problematic. It seems like the simplest fix would just be to ensure that this insertion is performed with an empty range, so it doesn’t try to delete anything.